### ISSUE TITLE
ICE with (hopefully) illegal CoerceUnsized impl

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#68013]
I was experimenting with `CoerceUnsized` and I encountered an ICE with [this example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0f18f9d5d6071f9a10586e87e5dfd0b0):

```rust
#[derive(Clone)]
struct Redirectable<'a, T: ?Sized> {
    data: Rc<Cell<&'a T>>
}

impl<U, T: CoerceUnsized<U>> CoerceUnsized<Redirectable<'_, U>> for Redirectable<'_, T> {}
```

This is *hopefully* an illegal `CoerceUnsized` impl, as it would permit us to do illegal mutation (see the full playground link above for the context). Possible duplicate of #44861.

Note that the ICE requires a feature gate.

**Comment by eddyb** [comment]
```
query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `&mut T: std::marker::Unsize`
#1 [coerce_unsized_info] processing `<Redirectable<'_, T> as std::ops::CoerceUnsized<Redirectable<'_, U>>>`
#2 [coherent_trait] coherence checking all impls of trait `std::ops::CoerceUnsized`
#3 [analysis] running analysis passes on this crate
end of query stack
```

Judging by this, I'd say this isn't that scary, since it ICEs while trying to check the validity of the `CoerceUnsized` impl.

But also, `&T` never implements `Unsize`, so this could never compile even if the checking was buggy, is `Rc<Cell<&T>>` the intended example, or is that one too many indirections?

**Comment by Alexendoo** [comment]
Fixed by #72194
